File: OBI -> Fase 2/2021/senha.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

int main(){
    ll n, m, k, p;
    ll comb, l, qnt, j = 0;

    string senha, tmp;

    cin >> n >> m >> k;

    vector<string> palavras;
    vector<char> letras;

    cin >> senha;

    for (ll i = 0; i < m; i++){
        cin >> tmp;
        palavras.push_back(tmp);
    }

    for (auto&& p: palavras){
        sort(p.begin(), p.end());
    }

    cin >> p;

    p -= 1;

    comb = pow(k, m);

    for (ll i = 1; i <= m; i++){
        l = p/(comb/pow(k, i));

        if (l >= k){
            qnt = l/k;
            l -= k * qnt;
        }

        letras.push_back((palavras[(i-1)][l]));
    }

    for (ll i = 0; i < size(senha); i++){
        if (senha[i] == '#'){
            cout << letras[j];
            j++;
        }
        else{
            cout << senha[i];
        }
    }

    cout << "\n";

    return 0;
}

File: OBI -> Fase 2/2021/senha.py
n, m, k = [int(i) for i in input().split()];
senha = input();

palavras = [];
letras = [];

j = 0;

for i in range(m):
    palavras.append("".join(sorted(input())));

p = int(input())
p -= 1;

comb = pow(k, m);

for i in range(1, m+1):
    l = p//(comb//pow(k, i));

    if (l >= k):
        qnt = l//k;
        l -= k * qnt;

    letras.append(palavras[(i-1)][l]);

for i in range(len(senha)):
    if (senha[i] == '#'):
        print(letras[j], end = "");
        j += 1;
    else:
        print(senha[i], end = "");

print();